iothub/device/src/Pipeline/MqttTransportHandler.cs

```diff
@@ -1108,18 +1140,18 @@ private async Task HandleReceivedCloudToDeviceMessageAsync(MqttApplicationMessag
                     if (Logging.IsEnabled)
                         Logging.Error(this, "Cannot 'reject' or 'abandon' a received message over MQTT. Message will be acknowledged as 'complete' instead.");
                 }
-
-                // Note that MQTT does not support Abandon or Reject, so always Complete by acknowledging the message like this.
-                try
-                {
-                    await receivedEventArgs.AcknowledgeAsync(CancellationToken.None).ConfigureAwait(false);
-                }
-                catch (Exception ex)
+            }
+            else if(!_isDeviceReceiveMessageCallbackSet)
+            {
+                _messageQueue.Enqueue(receivedCloudToDeviceMessage);
+                while(_messageQueue.Count > MessageQueueSize)
                 {
-                    // This likely happened because the connection was lost. The service will re-send this message so the user
-                    // can acknowledge it on the new connection.
+                    _messageQueue.TryDequeue(out _);
                     if (Logging.IsEnabled)
-                        Logging.Error(this, $"Failed to send the acknowledgement for a received cloud to device message {ex}"); ;
+                    {
+                        Logging.Info(this, $"Queue size of {MessageQueueSize} for C2D messages has been reached, removing oldest queued C2D message. " +
+                            $"To avoid losing further messages, set SetIncomingMessageCallbackAsync() to process the messages.");
```

Comment:
Add this into doc comments. Users don't usually look at logs until something really goes wrong.